sage/group_prover.sage

```diff
@@ -222,12 +222,12 @@ def prove_nonzero(R, exprs, assume):
       return (False, [exprs[expr]])
   allexprs = reduce(lambda a,b: numerator(a)*numerator(b), exprs, 1)
   for (f, n) in allexprs.factor():
-    if f not in nonzero:
+    if f not in nonzero and -f not in nonzero:
```

Comment:
Both  the elements of `nonzero` and of `allexprs.factor()` are prime factors obtained by factorization, so I think they can differ at most by their `unit()` (https://doc.sagemath.org/html/en/reference/structure/sage/structure/factorization.html#sage.structure.factorization.Factorization.unit), which is -1 or 1.

So an element of `nonzero` or `allexprs.factor()` cannot have an additional factor c with c != -1 and  c != 1 because then it wouldn't be a prime factor. But there could be a constant prime element c in `allexprs.factor()`, which is then trivially nonzero but the prover fails this.

So if we want to allow for this, I'd suggest (which does not on the `%` operator for which I can't find docs on):
```
  for (f, n) in allexprs.factor():
    if f not in nonzero and -f not in nonzero and not f.is_constant():
      ok = False
```

Docs are here:
https://doc.sagemath.org/html/en/reference/polynomial_rings/sage/rings/polynomial/multi_polynomial_element.html#sage.rings.polynomial.multi_polynomial_element.MPolynomial_polydict.is_constant

Reply:
> Are you sure? These are (multivariate) polynomial rings, and all constant terms are units for that.

I'm not sure at all.

> Specifically, I'm concerned about the case where you'd have factors a + 2*b or b + 1/2*a, depending on which variable is considered the one to make monic.

Hm yeah, I'd imagine this is the same everytime in the same ring class in sage but I don't know.

Would you then change every similar if within this function?

Reply:
> Hm yeah, I'd imagine this is the same everytime in the same ring class in sage but I don't know.

Indeed, but the same reasoning should IMO apply to the just the factor 1 or -1. If there is a canonical ordering of factors, and e.g. factor() always returns the lexicographically smallest list of factors, there should never be any ambiguity at all. The fact that the sign differs makes me suspicious, so I'd rather be overly cautious (the code still works fine with that change).

> Would you then change every similar if within this function?

Yeah. Perhaps introduce a `constant_ratio_between(a, b)` function or so to abstract it and make it a bit more readable?

Reply:
Ok I played around and I'm 99% convinced that 1 and -1 are complete. The reason is that the unit part is ignored when we convert the `Factorization` to a list in our code (one of the first examples in https://doc.sagemath.org/html/en/reference/structure/sage/structure/factorization.html), which then gives a canonical representation. 

Demonstration: 
(Sorry now even sage 9.5... Arch Linux is pretty quick with its updates...)


```
┌────────────────────────────────────────────────────────────────────┐
│ SageMath version 9.5, Release Date: 2022-01-30                     │
│ Using Python 3.10.2. Type "help()" for help.                       │
└────────────────────────────────────────────────────────────────────┘
sage: R.<ax,bx,ay,by,Az,Bz,Ai,Bi> = PolynomialRing(QQ,8,order='invlex')
sage: R
Multivariate Polynomial Ring in ax, bx, ay, by, Az, Bz, Ai, Bi over Rational Field

sage: R(10 * ax + bx / 2).factor()
(1/2) * (bx + 20*ax)
sage: R(20 * ax + bx).factor()
bx + 20*ax

sage: R(20 * ax + bx).factor().unit()
1
sage: R(10 * ax + bx/2).factor().unit()
1/2

sage: list(R(20 * ax + bx).factor())
[(bx + 20*ax, 1)]
sage: list(R(10 * ax + bx / 2).factor())
[(bx + 20*ax, 1)]
```

Why do 1 and -1 matter then at all, you may ask?
```
sage: R((-2 * (bx - ax)) ^ 1).factor()
(-2) * (bx - ax)
sage: R((-2 * (bx - ax)) ^ 2).factor()
(4) * (-bx + ax)^2
sage: R((-2 * (bx - ax)) ^ 3).factor()
(8) * (-bx + ax)^3
```
So it's not very consistent where they put the minus, not even consistent among odd exponents.

And indeed, I debugged the code on master and we hit `(bx - ax)^34` in `allexprs`, which then gets a sign change. 

For curiosity, can you check what this code does on your Sage 9.2? 

So with all that said, I'm in favor of sticking to current solution in this PR.

Reply:
I did some independent experiments and the only factorizations I was able to get Sage (9.4) to produce had the coefficients of the non-unit factors be integers with gcd 1. This would make them unique if it weren't for this +,- inconsistency.

  I considered that this is a regression in Sage >9.2 with `order="invlex"`, because replacing it with `order="lex"` makes the script work without the `-f not in nonzero` fix. But no.
  ```
  sage: a, b, c = QQ["a,b,c"].gens()
  ....: f = 1 / 2 * a - b
  ....: f.factor()
  ....: 
  (1/2) * (a - 2*b)
  sage: f2 = (1 / 2 * a - b) * (a - c)
  ....: f2.factor()
  (1/2) * (-a + c) * (-a + 2*b)
  ```

  I'd be fine with leaving the PR as is. Another solution would be to wrap the `factor()` method such that it behaves as we would expect, i.e. the smallest variable name in a factor gets a positive coefficient (and our other assumptions are asserted).

Reply:
nit: perhaps note in the commit description that this commit is required to make the script pass with sage 9.4.

Reply:
```
$ sage
┌────────────────────────────────────────────────────────────────────┐
│ SageMath version 9.2, Release Date: 2020-10-24                     │
│ Using Python 3.9.7. Type "help()" for help.                        │
└────────────────────────────────────────────────────────────────────┘
sage: R.<ax,bx,ay,by,Az,Bz,Ai,Bi> = PolynomialRing(QQ,8,order='invlex')
sage: R
Multivariate Polynomial Ring in ax, bx, ay, by, Az, Bz, Ai, Bi over Rational Field
sage: R(10 * ax + bx / 2).factor()
(1/2) * (bx + 20*ax)
sage: R(20 * ax + bx).factor()
bx + 20*ax
sage: R(20 * ax + bx).factor().unit()
1
sage: R(10 * ax + bx/2).factor().unit()
1/2
sage: list(R(20 * ax + bx).factor())
[(bx + 20*ax, 1)]
sage: list(R(10 * ax + bx / 2).factor())
[(bx + 20*ax, 1)]
sage: R((-2 * (bx - ax)) ^ 1).factor()
(2) * (-bx + ax)
sage: R((-2 * (bx - ax)) ^ 2).factor()
(4) * (-bx + ax)^2
sage: R((-2 * (bx - ax)) ^ 3).factor()
(8) * (-bx + ax)^3
sage: R((1/10 * ax + 1/15*bx)*(ax/7+bx)).factor()
(1/210) * (2*bx + 3*ax) * (7*bx + ax)
```

Reply:
Is this `(f not in nonzero) and (-f not in nonzero)` needed in other places where this pattern occurs below?

Reply:
> Another solution would be to wrap the factor() method such that it behaves as we would expect, i.e. the smallest variable name in a factor gets a positive coefficient (and our other assumptions are asserted).

That's a clean idea. I did this.

> Is this `(f not in nonzero) and (-f not in nonzero)` needed in other places where this pattern occurs below?

Not for the current group laws but I added the wrapper to the other places where we call `factor()`.